go-controller/pkg/networkmanager/nad_controller.go

```diff
@@ -180,20 +188,75 @@ func (c *nadController) Stop() {
 	c.networkController.Stop()
 }
 
-// RegisterNADHandler adds functions to be executed during NAD delete/update/add calls
-// usage of this function should be restricted to lightweight, non-blocking operations
-func (c *nadController) RegisterNADHandler(handler handlerFunc) error {
+// RegisterNADHandler adds functions to be executed during NAD delete/update/add calls.
+// usage of this function should be restricted to lightweight, non-blocking operations.
+// An optional needsUpdate predicate can be provided to filter update events for this handler;
+// if nil, all updates are delivered.
+// Returns an opaque handler ID that can be used to deregister the handler.
+func (c *nadController) RegisterNADHandler(handler handlerFunc, needsUpdate func(old, new *nettypes.NetworkAttachmentDefinition) bool) (NADHandlerID, error) {
+	c.Lock()
+	defer c.Unlock()
+	c.nextHandlerID++
+	id := c.nextHandlerID
+	c.handlers[id] = nadHandler{handler: handler, needsUpdate: needsUpdate}
+	return id, nil
+}
+
+// DeRegisterNADHandler removes a previously registered handler.
+func (c *nadController) DeRegisterNADHandler(id NADHandlerID) error {
 	c.Lock()
 	defer c.Unlock()
-	c.handlerFuncs = append(c.handlerFuncs, handler)
+	if _, ok := c.handlers[id]; !ok {
+		return fmt.Errorf("handler %d not found", id)
+	}
+	delete(c.handlers, id)
 	return nil
 }
 
-// executeHandlers should always be done under lock
+// executeHandlers runs handlers that were added by RegisterNADHandler.
+// Must be called with nadController locked.
 func (c *nadController) executeHandlers(nadName string, info util.NetInfo, removed bool) {
-	for _, handler := range c.handlerFuncs {
-		handler(nadName, info, removed)
+	// delete, no need to filter handler notification
+	if removed {
+		for _, h := range c.handlers {
+			h.handler(nadName, info, removed)
+		}
+		return
+	}
+	// add/update handlers should be filtered
+	idSet, ok := c.handlerUpdateDecisions[nadName]
+	if ok {
+		for _, id := range idSet.UnsortedList() {
+			if id == NetworkControllerUpdateID {
+				continue
+			}
+			if h, exists := c.handlers[id]; exists {
+				h.handler(nadName, info, removed)
+			}
+		}
+		// reset handlers after notifying all
+		delete(c.handlerUpdateDecisions, nadName)
+		return
+	}
+	// add/update with no filtering, send all
+	for _, h := range c.handlers {
+		h.handler(nadName, info, removed)
+	}
+}
```

Comment:
why no nad name means update all? shouldn't it mean nothing to update?

Reply:
it was a fallback, because syncAll does not addDecisions. I'll fix the comment